pkg/templating/api.go

```diff
@@ -153,6 +167,72 @@ func (lo ParentLabelSetAdder) Patch(cr resource.ParentResource, list []resource.
 	return list, nil
 }
 
+// NewAPIOrderedDeleter returns a new *APIOrderedDeleter.
+func NewAPIOrderedDeleter(c client.Client) *APIOrderedDeleter {
+	return &APIOrderedDeleter{kube: c}
+}
+
+// APIOrderedDeleter deletes the child resources in an order that is determined
+// by their priority noted in the child resource annotation. The child resources
+// with higher priority will be deleted first and their deletion will block
+// the lower priority ones.
+type APIOrderedDeleter struct {
+	kube client.Client
+}
+
+// Delete executes an ordered deletion of child resources depending on their
+// deletion priority.
+func (d *APIOrderedDeleter) Delete(ctx context.Context, list []resource.ChildResource) ([]resource.ChildResource, error) {
+	hp := minInt
+	del := []resource.ChildResource{}
+	for _, res := range list {
+		val, ok := res.GetAnnotations()[DeletionPriorityAnnotationKey]
+		// The zero-value sets a default but it doesn't necessarily mean that the
+		// resources with no annotation will be deleted last as user may want to
+		// mark some resources as last-to-be-deleted by giving them negative
+		// priority.
+		if !ok {
+			val = DeletionPriorityAnnotationZeroValue
+		}
+		p, err := strconv.Atoi(val)
+		if err != nil {
+			return nil, errors.Wrap(err, errPriorityToInt)
+		}
+		// We directly skip the case where we are sure that the resource
+		// has a lower priority than current highest priority regardless of
+		// whether it exists or not.
+		if p < hp {
+			continue
+		}
+		nn := types.NamespacedName{Name: res.GetName(), Namespace: res.GetNamespace()}
+		if err := d.kube.Get(ctx, nn, res); err != nil {
+			// The resources that do not exist anymore should not have any
+			// effect in our calculations.
+			if kerrors.IsNotFound(err) {
+				continue
+			}
+			return nil, errors.Wrap(err, errGetChildResource)
+		}
+		// If the priority of the resource is higher than our current highest
+		// priority level, then we should discard everything in the deletion queue
+		// and set the new highest priority.
+		if p > hp {
+			hp = p
+			del = []resource.ChildResource{res}
+			continue
+		}
+		// If the resource is on the highest priority level, then it should be
+		// deleted in this iteration.
+		del = append(del, res)
+	}
+	for _, res := range del {
+		if err := d.kube.Delete(ctx, res); client.IgnoreNotFound(err) != nil {
```

Comment:
Does this call block until the deleted resource's finalizers are gone?

Reply:
No, it doesn't block. If the resource still exists, that priority level would be still the highest one in the next reconcile.

Reply:
If we wanted it to block I believe we could do foreground deletion here. I guess the tradeoff would be that we might block until reconcile times out, but we would also reduce the number of reconciles because we wouldn't keep trying to delete every second if it was taking a resource a little while to get cleaned up. In the case of reconciliation time out we would be requeued anyway. In practice, the behavior to the end user would look the same I believe.

Reply:
I'd not like to block the process in the reconcile and timeout in an expected scenario. The processing of event queue would be slower as there are other objects and this reconciler doesn't utilize concurrency for processing the queue of CRs.